File: packages/php/src/CalendarError.php
<?php

namespace UmmAlQura;

class CalendarError extends \Exception
{
    private ErrorCode $errorCode;
    private ?string $field;
    private mixed $value;

    public function __construct(ErrorCode $errorCode, string $message, ?string $field = null, mixed $value = null)
    {
        parent::__construct($message);
        $this->errorCode = $errorCode;
        $this->field = $field;
        $this->value = $value;
    }

    public function getErrorCode(): ErrorCode
    {
        return $this->errorCode;
    }

    public function getField(): ?string
    {
        return $this->field;
    }

    public function getValue(): mixed
    {
        return $this->value;
    }

    public function toArray(): array
    {
        $result = [
            'error_code' => $this->errorCode->value,
            'message' => $this->getMessage(),
        ];
        if ($this->field !== null) {
            $result['field'] = $this->field;
        }
        if ($this->value !== null) {
            $result['value'] = $this->value;
        }
        return $result;
    }

    public function toJson(): array
    {
        return $this->toArray();
    }
}

File: packages/java/src/main/java/com/ummalqura/CalendarError.java
package com.ummalqura;

/**
 * Structured error response
 */
public class CalendarError extends RuntimeException {
    private final ErrorCode errorCode;
    private final String field;
    private final Object value;

    public CalendarError(ErrorCode errorCode, String message) {
        this(errorCode, message, null, null);
    }

    public CalendarError(ErrorCode errorCode, String message, String field, Object value) {
        super(message);
        this.errorCode = errorCode;
        this.field = field;
        this.value = value;
    }

    public ErrorCode getErrorCode() {
        return errorCode;
    }

    public String getField() {
        return field;
    }

    public Object getValue() {
        return value;
    }

    public String toJson() {
        StringBuilder sb = new StringBuilder();
        sb.append("{\"error_code\":\"").append(errorCode.getValue()).append("\"");
        sb.append(",\"message\":\"").append(getMessage()).append("\"");
        if (field != null) {
            sb.append(",\"field\":\"").append(field).append("\"");
        }
        if (value != null) {
            sb.append(",\"value\":").append(value);
        }
        sb.append("}");
        return sb.toString();
    }
}
